source/opt/instruction.cpp

```diff
@@ -109,5 +110,166 @@ void Instruction::ReplaceOperands(const std::vector<Operand>& new_operands) {
   operands_.insert(operands_.begin(), new_operands.begin(), new_operands.end());
   operands_.shrink_to_fit();
 }
+
+bool Instruction::IsReadOnlyLoad(IRContext* ctx) const {
+  if (IsLoad()) {
+    ir::Instruction* address_def = GetBaseAddress(ctx);
+    if (address_def->opcode() == SpvOpLoad ||
+        address_def->opcode() == SpvOpFunctionParameter) {
+      return false;
+    }
+    assert(address_def->opcode() == SpvOpVariable &&
+           "Expecting to get a variable.");
+
+    return address_def->IsReadOnlyVariable(ctx);
+  }
+  return false;
+}
+
+Instruction* Instruction::GetBaseAddress(IRContext* ctx) const {
+  assert(IsLoad() &&
+         "GetBaseAddress should only be called on load instructions.");
+  uint32_t base = GetSingleWordInOperand(0);
+  ir::Instruction* base_inst = ctx->get_def_use_mgr()->GetDef(base);
+  bool done = false;
+  while (!done) {
+    switch (base_inst->opcode()) {
+      case SpvOpAccessChain:
+      case SpvOpInBoundsAccessChain:
+      case SpvOpPtrAccessChain:
+      case SpvOpInBoundsPtrAccessChain:
+      case SpvOpImageTexelPointer:
+        base = base_inst->GetSingleWordInOperand(0);
+        base_inst = ctx->get_def_use_mgr()->GetDef(base);
+        break;
+      default:
+        done = true;
+        break;
+    }
+  }
+  return base_inst;
+}
+
+bool Instruction::IsReadOnlyVariable(IRContext* ctx) const {
+  assert(ctx->module()->HasCapability(SpvCapabilityShader) &&
+         "We don't know how enough about non-shaders to answer this question.");
+  uint32_t storage_class = GetSingleWordInOperand(0);
+  ir::Instruction* type_def = ctx->get_def_use_mgr()->GetDef(type_id());
+
+  switch (storage_class) {
+    case SpvStorageClassUniformConstant:
+      if (!type_def->IsVulkanStorageImage(ctx) &&
+          !type_def->IsVulkanStorageTexelBuffer(ctx)) {
+        return true;
+      }
+      break;
+    case SpvStorageClassUniform:
+      if (!type_def->IsVulkanStorageBuffer(ctx)) {
+        return true;
+      }
+      break;
+    case SpvStorageClassPushConstant:
+    case SpvStorageClassInput:
+      return true;
+    default:
+      break;
+  }
+
+  auto decorations = ctx->get_decoration_mgr()->GetDecorationsFor(
+      result_id(), SpvDecorationNonWritable);
+  return !decorations.empty();
+}
+
+bool Instruction::IsVulkanStorageImage(IRContext* ctx) const {
+  if (opcode() == SpvOpTypePointer) {
+    ir::Instruction* base_type =
+        ctx->get_def_use_mgr()->GetDef(GetSingleWordInOperand(1));
+    return base_type->IsVulkanStorageImage(ctx);
+  }
+
+  if (opcode() != SpvOpTypeImage) {
+    return false;
+  }
+
+  if (GetSingleWordInOperand(0) == SpvDimBuffer) {
+    return false;
+  }
+
+  // Check if the image is sampled.  If we do not know for sure that it is,
+  // then assume it is a storage image.
+  return GetSingleWordInOperand(6) != 1;
```

Comment:
Use a const int instead of hardcoding 6?

Reply:
Done, for this and other constant indices for operands.